### PR TITLE
Vqa speed improvements

### PR DESCRIPTION
**Pull request by klshuster**
1. Adds a VQA Dataset class for use in the PytorchDataLoader to improve VQA training times.
2. Generalizes PytorchDataTeacher to allow alternative datasets for tasks and different batch preparation functions (`collate_fn`) for models
3. Minor fixes to `build_pytorch_data`, and turns off cache in `image_featurizers`
4. Changes the way that the PytorchDataLoader handles epochs - now, only one iterator over the data is created, compared to a new `iter(self.data)` for every teacher

### CODE REVIEW

File: parlai/core/agents.py [hunk 1]
```diff
@@ -284,7 +284,10 @@ def load_agent_module(opt):
         return None
 
 def get_agent_module(dir_name):
-    if ':' in dir_name:
+    if '.' in dir_name:
+        module_name = dir_name.split(':')[0]
+        class_name = dir_name.split(':')[-1]
+    elif ':' in dir_name:
```
Comment (alexholdenmiller) [inline parlai/core/agents.py]: what broke here?

is `len(dir_name.split(':')) > 1` when `'.' in dir_name`?

File: parlai/core/pytorch_data_teacher.py [hunk 2]
```diff
@@ -326,7 +325,7 @@ def __getitem__(self, index):
                 return (index, ep)
 
     def __len__(self):
-        return self.num_eps
+        return int(self.num_episodes() * max(self.opt.get('num_epochs', 1), 1))
```
Comment (alexholdenmiller) [inline parlai/core/pytorch_data_teacher.py]: what's this for?

File: parlai/core/pytorch_data_teacher.py [hunk 3]
```diff
@@ -385,31 +386,57 @@ def __init__(self, opt, shared=None):
...
             self.dataset = shared['dataset']
             self.pytorch_dataloader = shared['pytorch_dataloader']
             self.lastYs = shared['lastYs']
+            self.data = shared['data']
 
         self.num_batches = math.ceil(self.dataset.num_episodes()/self.bsz)
         self.reset()
 
+    def get_dataset_class(self, opt):
+        dataset_name = opt.get('dataset')
```
Comment (alexholdenmiller) [inline parlai/core/pytorch_data_teacher.py]: Can you add a docstring here? Good to provide a few examples of how to specify the dataset name

**Comment by alexholdenmiller** [comment]
Nice though! Can you post the speedup here?

File: parlai/core/agents.py (hunk 1 above)
Comment (klshuster) [reply parlai/core/agents.py]: I tried using an agent from parlai_external, which i had to specify as `-m "parlai_external.agents.mlb.mlb:MlbAgent"`

File: parlai/core/pytorch_data_teacher.py (hunk 2 above)
Comment (klshuster) [reply parlai/core/pytorch_data_teacher.py]: This is intended to fix the issue that a new data-loading iterator was constructed for each teacher in the `reset_data` function; now, only one iterator is created, and does not stop iterating until the appropriate amount of examples have been returned. I will change this so that it works for validation/test (as right now it looks like it would do too many iterations)

File: parlai/core/pytorch_data_teacher.py (hunk 3 above)
Comment (klshuster) [reply parlai/core/pytorch_data_teacher.py]: sure!

File: parlai/core/agents.py (hunk 1 above)
Comment (alexholdenmiller) [reply parlai/core/agents.py]: shouldn't the module/class-name code below work for that? it looks like in both cases you should get `module_name = parlai_external.agents.mlb.mlb` and `class_name = MlbAgent`, right?

File: parlai/core/pytorch_data_teacher.py [hunk 4]
```diff
@@ -418,11 +424,25 @@ def __init__(self, opt, shared=None):
...
+        """ To use a custom dataset (as opposed to the StreamDataset above),
+            you can subclass the pytorch Dataset class in a task's ``agents.py``
+            file, and then specify it in the command line.
+
+            For example, the VQA v1 task provides a custom dataset, which can
+            be specified on the command line as follows:
+            ``--dataset parlai.tasks.vqa_v1.agents:VQADataset``
+
+            Where VQADataset is the dataset defined in the vqa_v1 teacher
+            file.
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/core/pytorch_data_teacher.py]: Sure, I will change that

File: parlai/core/agents.py (hunk 1 above)
Comment (klshuster) [reply parlai/core/agents.py]: You know, I could've sworn this wasn't working for me before, but it does in fact work without my addition. Will remove in the next commit

**Comment by alexholdenmiller** [comment]
let's get those speed differences explained next!

**Comment by klshuster** [comment]
New commit adds additional speedup, to 1250 ex/sec (brings it to roughly 2x slower than pytorch)

**Comment by jaseweston** [comment]
yay!

**Comment by klshuster** [comment]
With this most recent commit, I can achieve ~2400 ex/sec, with only 1 GPU and around 9 GB of RAM (about 12% slower than pytorch)

**Comment by jaseweston** [comment]
great!

**Comment by jaseweston** [comment]
how is test acc?

**Comment by klshuster** [comment]
there was an issue with my machine gpu, so had to switch machines/extract the features again; will have results once that's taken care of

**Comment by klshuster** [comment]
Validation set accuracy for our implementation of the model is in line with (and at times superior to) validation accuracy achieved by the pytorch implementation of the model - after training for 100 epochs with the same model configurations, our model had 57.5% validation accuracy, while the pytorch implementation had 54.1% accuracy. I will be pushing a final commit to clean up some extraneous changes that made their way into this branch and will merge after

**Comment by alexholdenmiller** [comment]
great!
